### PR TITLE
add: Generating DAG from CUE program

### LINKED ISSUE #9: [Feature]: Create a Dag from a Cue program
### Description

Create a dag from a cue program :rocket:

Here is the schema :
- With Cue API, you get informations about the given Cue program, as JSON
- With you custom Go Dag librairy, create a Dag from this JSON

### Expected behaviour

Here is what you have to do :

- [ ] Update the documentation and the README.md
- [ ] Implement the feature
- [ ] Add unit-tests and functional tests
- [ ] Dockerize the project
- [ ] Ask Tom and Damien to review your code

### Environment

/

### Proposed solution

/

### PR DESCRIPTION
**Pull request by EdenComp**
# Description

This PR adds a relation between the DAG creation library and the CUE program parser.
We can now get a generated DAG from the content of the CUE program passed as parameter.

# Changes include

- [X] Bugfix (non-breaking change that solves an issue)
- [X] New feature (non-breaking change that adds functionality)
- [X] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

main.go has been modified

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the needed labels
- [x] I have linked this PR to an issue
- [x] I have linked this PR to a milestone
- [x] I have linked this PR to a project
- [x] I have tested this code
- [x] I have added / updated tests (unit / functionals / end-to-end / ...)
- [x] I have updated the README and other relevant documents (guides...)
- [x] I have added sufficient documentation both in code, as well as in the READMEs

# Additional comments

Bugfix: DAG printer issues (recursive printing and format)

### CODE REVIEW

File: dag/struct-root.go [hunk 1]
```diff
@@ -12,7 +12,8 @@ type Root struct {
 func (r *Root) PrintDag(i interface{}) {
 	fmt.Println("Root:", r.Name, "\n")
 	for _, m := range r.Members {
-		m.GetValue(i)
+		m.GetValue(i, 0)
+		fmt.Println("")
```
Comment (TomChv) [inline dag/struct-root.go:16]: What is the purpose of this print?

File: cuetojson/values.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+func AppendValuesToDag(root *dag.Root, programs []CueProgram) {
+	for _, program := range programs {
+		addNode(root, nil, program.Values, nil)
+	}
+	root.PrintDag(" --> ")
+}
+
+func addNode(root *dag.Root, node *dag.Node, value *cue.Value, prev *cue.Iterator) {
+	iterator, err := value.Fields()
```
Comment (TomChv) [inline cuetojson/values.go]: Why you didn't used function [`Walk`](https://cuetorials.com/go-api/basics/traversing/#walking-a-value) to explore value?

File: cue/ValidActions.cue [hunk 3]
```diff
@@ -1,28 +1,38 @@
...
+		"runs-on": "ubuntu-latest"
+		"container": {
+			"image": "string"
+		}
+		"needs": [
+			"job1",
+			"job2",
+		]
+		#steps
+	}
```
Comment (TomChv) [inline cue/ValidActions.cue:37]: What is this? It's not a valid GitHub action schema

**Review by TomChv — COMMENTED** [review]
That's a lot of work, I left few questions.

Global comment:
- Do not `merge` commit from upstream, rebase instead to avoid ugly `merge branch x into x` commit message in your PR
- Add concrete example, I do not really understand what your PR do, does it work?
- Your GitHub Action schema is false, what was the original purpose?

File: dag/struct-root.go (hunk 1 above)
Comment (Doozers) [reply dag/struct-root.go:16]: It print the created DAG based on the cue program passed as argument

File: cue/ValidActions.cue (hunk 3 above)
Comment (Doozers) [reply cue/ValidActions.cue:37]: Yes but we tried to retrieve any info from the original CUE program that validate github actions but it seems like it is impossible to get it if the fields are requirements and not values.

Comment (Doozers) [reply cue/ValidActions.cue:37]: Screenshot from 2022-05-22 21-28-37

Comment (Doozers) [reply cue/ValidActions.cue:37]: Screenshot from 2022-05-22 21-28-37

Comment (Doozers) [reply cue/ValidActions.cue:37]: Screenshot from 2022-05-22 21-26-39

File: dag/struct-root.go (hunk 1 above)
Comment (Doozers) [reply dag/struct-root.go:16]: Screenshot from 2022-05-22 21-50-28

File: cue/ValidActions.cue (hunk 3 above)
Comment (Doozers) [reply cue/ValidActions.cue:37]: new-screen

File: dag/struct-root.go (hunk 1 above)
Comment (TomChv) [reply dag/struct-root.go:16]: Awesome!

File: cue/ValidActions.cue (hunk 3 above)
Comment (TomChv) [reply cue/ValidActions.cue:37]: Yeah it's normal, you must retrieve/read a file that implement the definition, not the definition itself haha

File: tests/dagFromCue_test.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	if root.Members[0].Links[1].Links[0].Value != "salut" {
+		t.Errorf("Expected salut, got %s", root.Members[0].Links[0].Value)
+	}
+
+	if root.Members[0].Links[1].Links[0].Links[0].Value != "bonsoir = au revoir" {
+		t.Errorf("Expected \"bonsoir = au revoir\", got %s", root.Members[0].Links[0].Value)
+	}
+
+	if root.Members[1].Value != "info = bonsoir" {
+		t.Errorf("Expected \"info = bonsoir\", got %s", root.Members[1].Value)
```
Comment (TomChv) [inline tests/dagFromCue_test.go]: You have t.Assert assert stuff for this :)

**Review by TomChv — COMMENTED** [review]
Thanks for your change,
I left a last comment

File: tests/dagFromCue_test.go (hunk 4 above)
Comment (Doozers) [reply tests/dagFromCue_test.go]: done :heavy_check_mark:

**Review by TomChv — APPROVED** [review]
LGTM

File: cuetojson/values.go (hunk 2 above)
Comment (EdenComp) [reply cuetojson/values.go]: The use of Walk will be a subject of the next sprint (thanks to the call we had this morning)

**Review by RezaRahemtola — CHANGES_REQUESTED** [review]
LGTM, you just need to :
- update unit tests
- add functional tests to make sure your binary works as expected
- add a Continuous Integration with GitHub Actions that builds the project and launches unit and functional tests on push/PR.

**Review by TomChv — APPROVED** [review]
LGTM, the code can definitely be cleaner but you could do this as a follow up :rocket:

**Review by RezaRahemtola — APPROVED** [review]

**Pull request merged**